File: r/math/collatz.R
# Print the start below 1000 with the longest Collatz sequence, picked by which.max over a vector of lengths.
# Run: Rscript collatz.R

limit <- 1000

chain_length <- function(start) {
  value <- start
  count <- 1
  while (value != 1) {
    value <- if (value %% 2 == 0) value %/% 2 else value * 3 + 1
    count <- count + 1
  }
  count
}

starts <- seq_len(limit - 1)
chain_lengths <- vapply(starts, chain_length, numeric(1))
longest <- which.max(chain_lengths)

cat(paste(c(starts[longest], chain_lengths[longest]), collapse = " "), "\n", sep = "")

File: swift/math/collatz.swift
// Print the start below 1000 with the longest Collatz sequence, picked by max(by:) over a range.
// Run: swift collatz.swift

let limit = 1000

func chainLength(_ start: Int) -> Int {
    var value = start
    var length = 1
    while value != 1 {
        value = value.isMultiple(of: 2) ? value / 2 : value * 3 + 1
        length += 1
    }
    return length
}

let longest = (1..<limit).max { chainLength($0) < chainLength($1) }!
print("\(longest) \(chainLength(longest))")
